JP-3799: First frame bright

karllark opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3799](https://jira.stsci.edu/browse/JP-3799)

<!-- describe the changes comprising this PR here -->
This PR modifies the firstframe step to optionally not flag the 1st group (aka frame) as DO_NOT_USE for pixels that have ramps that saturation between the 2nd and 3rd groups.   In the case of a pixel having a large signal, it has been found that the slope estimated from the 2nd-1st groups and 3rd-2nd groups is equivalent within 1% indicating that the firstframe effect is small compared to the signal in this specific case. This means that for pixels seeing very bright signal (i.e., saturating between the 2nd and 3rd groups), that the slope can be measured from the 1st and 2nd groups. This provides a slope measurement where previously there was none (i.e., a NaN value).

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

jemorrison commented on docs/jwst/firstframe/arguments.rst:
```diff
@@ -1,4 +1,8 @@
+
+``--bright_use_group1`` (boolean, default=True)
+    If True, for pixels that have the saturation flag set for group 3, 
+    but not group 2 will not have their group 1 groupdq set to DO_NOT_USE.
```
Suggestion (some like): 
If True, for pixels that have the saturation flag set for group 3,  but not group 2  then the groupdq for group 1 will not be set DO_NOT_USE. This will allow a slope to be determined for this pixel.

karllark replied:
I've reworded motivated by your comment.  Hopefully the new wording works.

jemorrison reviewed (approved):
Code looks good and the test added checks that it is working.

melanieclarke commented on docs/jwst/firstframe/arguments.rst:
```diff
@@ -1,4 +1,9 @@
+``--bright_use_group1`` (boolean, default=False)
+    If True, setting the group 1 groupdq to DO_NOT_USE will not be done 
+    for pixels that have the saturation flag set for group 3, but not group 2.
+    This will allow a slope to be determined for this pixel.
```
It might be helpful to add a note here about when this is appropriate (i.e. for very bright targets).

karllark replied:
Will do.  It could be used for *all* data, just will only have an impact for those pixels with really bright sources.  For all other pixels, the firstframe flagging will happen just as it current does.  Hopefully this is clear, but maybe I should say this explicitly?  Please advise.

melanieclarke replied:
Yes, I think it would help to be explicit here - it was a little surprising to me at first that turning it on made no difference when I tested it on data with no very bright sources.

melanieclarke reviewed (commented):
Looks good! One quick suggestion for the docs for the new argument.

I'll start regression tests.

melanieclarke commented:
Regression tests running here:
https://github.com/spacetelescope/RegressionTests/actions/runs/11898230791

karllark commented:
> It might be useful to expand this to unflag group 1 in the cases where saturation occurs in group 2 as well, not just between groups 2 and 3. This would allow for potential future support of single-group data (e.g., using the suppress_one_group=False approach) without having to disable the first-frame step entirely.

I agree this is a good idea.  I will update the PR to make this change.  It is a straightforward change.

karllark commented:
I updated this PR based on PR comments and the recent David Law comment.  Based on an in-person conversation, we both agree that the minor change to the flagging made no change to the main goal of this PR and supports another use case.

I rebased to get to get up-to-date with the main branch.   (First mistakenly did this to the "master" branch as my fork still has "master" - need to fix this, and then did the rebase to main.  Hopefully this fixed things  and didn't mess anything up.).

So, I think this PR is now ready for merging.

melanieclarke reviewed (commented):
Thanks for the update - I'll re-run regression tests.

It looks like some of the docs are still referring to saturating between 2 and 3 - should those be updated, since group 2 can also now be saturated?

karllark commented:
> It looks like some of the docs are still referring to saturating between 2 and 3 - should those be updated, since group 2 can also now be saturated?

Good catch.  I've (hopefully) updated things in all the right places.

melanieclarke commented on docs/jwst/firstframe/arguments.rst:
```diff
@@ -1,4 +1,14 @@
+    If True, setting the group 1 groupdq to DO_NOT_USE will not be done 
+    for pixels that have the saturation flag set for group 3.  
+    This will allow a slope to be determined for pixels that saturate
+    between groups 2 and 3.
```
```suggestion
    in group 2 or 3.
```

karllark replied:
Argh.  Missed that one.

melanieclarke reviewed (approved):
I found one last place that was referring to saturating between group 2 and 3 - suggestion below.

I reran regression tests at the same link above.  There are some unrelated failures, but nothing related to this PR.

I'll approve now from my perspective.  @drlaw1558 - are you happy with this one now?

drlaw1558 reviewed (approved):
LG2M

karllark commented:
I've fixed the last doc issue and pushed the commit.  Seems to be taking sometime for github to process this update.  Just a FYI.

melanieclarke commented:
@karllark - I still don't see the update to the docs.  Can you check that you pushed all the changes up?

karllark commented:
I did.  At the top of this PR you can see a wheel going around saying there has been an update, just hasn't been ingested somehow yet.

karllark commented:
And on my fork/branch the commit is present.

melanieclarke commented:
Okay, thanks.  I'll close and reopen and see if that triggers an update.

melanieclarke commented:
All good now!  I will merge.
